// @companion-module/base/dist/internal/feedback.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.callFeedbackOnDefinition = exports.convertFeedbackInstanceToEvent = void 0;
function convertFeedbackInstanceToEvent(type, feedback) {
    return {
        type: type,
        id: feedback.id,
        feedbackId: feedback.feedbackId,
        controlId: feedback.controlId,
        options: feedback.options,
    };
}
exports.convertFeedbackInstanceToEvent = convertFeedbackInstanceToEvent;
function callFeedbackOnDefinition(definition, feedback) {
    if (definition.type === 'boolean') {
        return definition.callback({
            ...convertFeedbackInstanceToEvent('boolean', feedback),
            type: 'boolean',
            _rawBank: feedback.rawBank,
        });
    }
    else {
        return definition.callback({
            ...convertFeedbackInstanceToEvent('advanced', feedback),
            type: 'advanced',
            image: feedback.image,
            _page: feedback.page,
            _bank: feedback.bank,
            _rawBank: feedback.rawBank,
        });
    }
}
exports.callFeedbackOnDefinition = callFeedbackOnDefinition;
